routing/src/interface.rs

```diff
@@ -0,0 +1,363 @@
+    // the ones to trigger the enforcement of the new state
+    // (e.g. by requesting the interface manager to bring up/down
+    // interfaces in kernel), or, instead, they just update the
+    // interface object seen by the rest of the routing system.
+
+    //////////////////////////////////////////////////////////////////
+    /// Attach an interface to a VRF. It is assumed that both the
+    /// interface and the VRF exist.
+    //////////////////////////////////////////////////////////////////
+    pub fn attach(&mut self, vrf: &Arc<Vrf>) -> Result<(), RouterError> {
+        if let Some(exist_vrf) = &self.vrf {
+            if !Arc::ptr_eq(exist_vrf, vrf) {
+                Err(RouterError::AlreadyAttached)
+            } else {
+                Ok(())
+            }
+        } else {
+            self.vrf = Some(vrf.clone());
+            Ok(())
+        }
+    }
+
+    //////////////////////////////////////////////////////////////////
+    /// Detach an interface from its VRF (whichever it is)
+    //////////////////////////////////////////////////////////////////
+    pub fn detach(&mut self) {
+        self.vrf.take();
+    }
+
+    //////////////////////////////////////////////////////////////////
+    /// Get the VRF that an interface is attached to, or None otherwise
+    //////////////////////////////////////////////////////////////////
+    pub fn get_vrf(&self) -> Option<&Vrf> {
+        self.vrf.as_deref()
+    }
+
+    //////////////////////////////////////////////////////////////////
+    /// Add (assign) an IP address to an interface
+    //////////////////////////////////////////////////////////////////
+    pub fn add_ifaddr(&mut self, ifaddr: &IfAddress) {
+        self.addresses.insert(*ifaddr);
+    }
+
+    //////////////////////////////////////////////////////////////////
+    /// Del (unassign) an IP address from an interface
+    //////////////////////////////////////////////////////////////////
+    pub fn del_ifaddr(&mut self, ifaddr: &IfAddress) {
+        self.addresses.remove(ifaddr);
+    }
+
+    //////////////////////////////////////////////////////////////////
+    /// Tell if an interface has a certain IP address assigned
+    /// (regardless of the mask)
+    //////////////////////////////////////////////////////////////////
+    pub fn has_address(&self, address: &IpAddr) -> bool {
+        for (addr, _) in &self.addresses {
+            if addr == address {
+                return true;
+            }
+        }
+        false
+    }
+}
+
+/// A table of network interface objects, keyed by some ifindex (u32)
+pub struct IfTable(HashMap<u32, Interface>);
+
+#[allow(dead_code)]
+impl IfTable {
+    //////////////////////////////////////////////////////////////////
+    /// Create an interface table. All interfaces should live here.
+    //////////////////////////////////////////////////////////////////
+    pub fn new() -> Self {
+        Self(HashMap::new())
+        // TODO: set a fast hasher
+    }
```

Comment:
Yup.  This is a fair todo.  Likely `ahash` but we should benchmark if it turns out to be a control here

Reply:
Glad that you mention `ahash` because that's the one I was thinking of using. Do you think it's a good choice?

Reply:
Every benchmark of `ahash` I have ever done shows that it is appallingly fast.  It feels like cheating.  It is a really high-quality hash in terms of collision rate as well